### PR TITLE
cmd/utils, crypto: add --lightkdf flag for lighter KDF

### PR DESCRIPTION
**Pull request by wildfjre**
Addresses

https://github.com/ethereum/go-ethereum/issues/1148
https://github.com/ethereum/go-ethereum/issues/1911

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Wed Oct 28 16:31:19 UTC 2015_

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -157,6 +157,10 @@ var (
 		Name:  "fast",
 		Usage: "Enables fast syncing through state downloads",
 	}
+	LightKDFFlag = cli.BoolFlag{
+		Name:  "lightkdf",
+		Usage: "Reduce KDF iterations trading memory usage vs security",
```
Comment (Gustav-Simonsson) [inline cmd/utils/flags.go]: Can be improved a bit with "Reduce KDF memory usage at some expense of KDF strength"

File: crypto/key_store_passphrase.go [hunk 2]
```diff
@@ -87,11 +103,12 @@ func (ks keyStorePassphrase) GetKeyAddresses() (addresses []common.Address, err
...
 	authArray := []byte(auth)
 	salt := randentropy.GetEntropyCSPRNG(32)
-	derivedKey, err := scrypt.Key(authArray, salt, scryptN, scryptr, scryptp, scryptdkLen)
+	now := time.Now()
+	derivedKey, err := scrypt.Key(authArray, salt, ks.scryptN, ks.scryptR, ks.scryptP, ks.scryptDKLen)
 	if err != nil {
 		return err
 	}
-
+	fmt.Println("took: ", time.Since(now))
```
Comment (fjl) [inline crypto/key_store_passphrase.go]: Please remove debug output.

Comment (Gustav-Simonsson) [inline crypto/key_store_passphrase.go]: Please remove as it was only used to benchmark.

File: crypto/key_store_passphrase.go [hunk 3]
```diff
@@ -45,19 +46,35 @@ import (
 
 const (
 	keyHeaderKDF = "scrypt"
-	// 2^18 / 8 / 1 uses 256MB memory and approx 1s CPU time on a modern CPU.
-	scryptN     = 1 << 18
-	scryptr     = 8
-	scryptp     = 1
-	scryptdkLen = 32
+	KDFStandard  = iota
+	KDFLight     = iota
```
Comment (fjl) [inline crypto/key_store_passphrase.go]: Style: Please put 'enum' constants in a separate block and add a comment, e.g.

```go
const (
    KDFStandard = iota
    KDFLight
)

const (
    // the other stuff
)
```

File: eth/backend.go [hunk 4]
```diff
@@ -111,6 +111,7 @@ type Config struct {
 	AutoDAG   bool
 	PowTest   bool
 	ExtraData []byte
+	LightKDF  bool
```
Comment (Gustav-Simonsson) [inline eth/backend.go]: This field is not actually used so can be removed.

File: common/natspec/natspec_e2e_test.go [hunk 5]
```diff
@@ -128,7 +128,7 @@ func testEth(t *testing.T) (ethereum *eth.Ethereum, err error) {
 	db, _ := ethdb.NewMemDatabase()
 	addr := common.HexToAddress(testAddress)
 	core.WriteGenesisBlockForTesting(db, core.GenesisAccount{addr, common.String2Big(testBalance)})
-	ks := crypto.NewKeyStorePassphrase(filepath.Join(tmp, "keystore"))
+	ks := crypto.NewKeyStorePassphrase(filepath.Join(tmp, "keystore"), crypto.StandardScryptN, crypto.StandardScryptP)
```
Comment (fjl) [inline common/natspec/natspec_e2e_test.go]: Please change to light.

File: crypto/crypto.go [hunk 6]
```diff
@@ -215,7 +215,7 @@ func Decrypt(prv *ecdsa.PrivateKey, ct []byte) ([]byte, error) {
 
 // Used only by block tests.
 func ImportBlockTestKey(privKeyBytes []byte) error {
-	ks := NewKeyStorePassphrase(common.DefaultDataDir() + "/keystore")
+	ks := NewKeyStorePassphrase(common.DefaultDataDir()+"/keystore", StandardScryptN, StandardScryptP)
```
Comment (fjl) [inline crypto/crypto.go]: Please change to light.

**Comment by fjl** [comment]
:+1:

**Comment by Gustav-Simonsson** [comment]
:+1:

**Pull request merged**